### PR TITLE
Add undo/redo/crop/scale/modify canvas buttons

### PR DESCRIPTION
**Pull request by precla**
Add undo/redo/crop/scale/modify canvas buttons when required. Since those are not always needed.
For example ksnip has those built-in.

### CODE REVIEW

**Comment by precla** [comment]
Ok, now this one lacks functionality when clicking on the buttons and the buttons are always visible.
Quite a lot of classes "above" this. So I'm not quite sure how to call, for example `CoreView::undoAction`, from within `TransformationTools::undoAction`. Can you assist with this?

**Comment by precla** [comment]
Screenshot:

**Comment by kevincolyer** [comment]
Thanks for working on this. Will be a great addition!

**Comment by precla** [comment]
Done, tested & working.
Test & code review please :)

**Comment by DamirPorobic** [comment]
I'll have look into it today, yesterday evening GitHub was done for me.  

One thing that I've already noticed, not wrong just might be misleading. The name "Transformation" might be used in item operations too, there would it be like rotate, scale and so on, that would also be Transformation. But to be honest, I've got no better name right now for it.

File: src/widgets/TransformationTools.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+Q_SIGNALS:
+	void undo() const;
+	void redo() const;
+	void crop() const;
+	void scale() const;
+	void modifyCanvas() const;
+
+private:
+	QActionGroup *mActionGroup;
```
Comment (DamirPorobic) [inline src/widgets/TransformationTools.h:48]: For what do you need this ActionGroup and ActionToButton? I don't see any usage except the assignment?

File: src/widgets/TransformationTools.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+void TransformationTools::modifyCanvasTriggered()
+{
+	emit modifyCanvas();
+}
+
+QAction *TransformationTools::createAction(const QString &tooltip, const QIcon &icon)
+{
+	auto action = new CustomToolButtonAction(this);
+	action->setCheckable(true);
```
[the remark this answers is not in the document]
Comment (DamirPorobic) [reply src/widgets/TransformationTools.cpp:88]: Or all of them stay selected 
image

File: src/gui/annotator/settings/AnnotationTransformationSelection.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	void crop() const;
+	void scale() const;
+	void modifyCanvas() const;
+
+private:
+	QBoxLayout *mMainLayout;
+	TransformationTools *mTransformationTools;
+
+private slots:
+	void transformationUndo();
```
Comment (DamirPorobic) [inline src/gui/annotator/settings/AnnotationTransformationSelection.h:49]: Can be const.

File: src/gui/annotator/AnnotationWidget.h [hunk 4]
```diff
@@ -82,6 +86,9 @@ Q_OBJECT
 private slots:
 	void persistDockWidgets();
 	void restoreDockWidgets();
+	void scaleTriggered();
```
Comment (DamirPorobic) [inline src/gui/annotator/AnnotationWidget.h:89]: Can be const.

**Review by DamirPorobic — COMMENTED** [review]
I'm missing an option to disable the Transformation Settings Widget explicitly. Or am I missing something?

**Comment by precla** [comment]
> I'm missing an option to disable the Transformation Settings Widget explicitly. Or am I missing something?
You can enable it with:
`kImageAnnotator->showTransformationSettingsWidget()`

**Comment by DamirPorobic** [comment]
> You can enable it with:
> kImageAnnotator->showTransformationSettingsWidget()

Strange, I see the method in the code but not in the Review on GitHub.

File: src/widgets/TransformationTools.h (hunk 1 above)
Comment (precla) [reply src/widgets/TransformationTools.h]: leftover from copy paste :D

**Comment by DamirPorobic** [comment]
One problem though, currently we setup the widgets and call the `restoreDockWidgets` to load the last saved position of widgets from config. Now the Transformation widget is added after the `restoreDockWidgets` call so it always ends up in the bottom right corner. Basically you cannot persist the position of the widget. Might be enough to just call the `restoreDockWidgets` method in the `showTransformationSettingsWidget` method after adding the widget.

**Comment by precla** [comment]
> One problem though, currently we setup the widgets and call the `restoreDockWidgets` to load the last saved position of widgets from config. Now the Transformation widget is added after the `restoreDockWidgets` call so it always ends up in the bottom right corner. Basically you cannot persist the position of the widget. Might be enough to just call the `restoreDockWidgets` method in the `showTransformationSettingsWidget` method after adding the widget.

Added, seems to do the trick.

All fine now? Except for the naming. `Transformation` doesn't really fit, since this includes undo/redo - which are not that much of a transformation.
Maybe `Modification`? You got any ideas?

**Comment by DamirPorobic** [comment]
`Modification` might be a bit better but then again, what are you modifying, could be anything. Problem is that the operations are mixed, undo/redo is an overall operation affecting items and image, scale and crop only affect the image and modify canvas effects only what's outside the image, there is nothing common except that they're available through other calls. No clue. Should we leave it as it is?

**Comment by precla** [comment]
Yeah, undo/redo is not a modification per se. Can't come up with something else atm.. Leave it as is and 'one day' when we come up with something different, change it?

**Comment by DamirPorobic** [comment]
Yeah, let's go with that. The only drawback is that the `showTransformationSettingsWidget` is part of our external interface so we cannot just like that rename it but yeah, the "one day" might be far away :)

You can squash it, I'll have a quick look through and test later today when I find time and merge it then.

**Comment by precla** [comment]
> Yeah, let's go with that. The only drawback is that the `showTransformationSettingsWidget` is part of our external interface so we cannot just like that rename it but yeah, the "one day" might be far away :)
> 
> You can squash it, I'll have a quick look through and test later today when I find time and merge it then.

Hm, maybe `General Selection` , since `General Settings` is already taken (and only has Zoom in it) ? Then we'd have `showGeneralSelectionWidget`. When I think about it, undo/redo/crop/... ain't settings, so renaming to selection should make sense?

a) `showGeneralSelectionWidget` (I'd prefer this one), and rename all of `TransformationSettings` into `GeneralSelection`

b) `showTransformationSelectionWidget`

**Comment by DamirPorobic** [comment]
Yeah, something `General` doesn't sound bad, it's not settings indeed but doesn't sound like selection either. Maybe `Controls`? `GeneralControlsWidget` or just `ControlsWidget`?

**Comment by precla** [comment]
`ControlsWidget` sounds great and is short. Will change now to that.

File: src/widgets/Controls.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+Q_SIGNALS:
+	void undo() const;
+	void redo() const;
+	void crop() const;
+	void scale() const;
+	void modifyCanvas() const;
+
+private:
+	QHBoxLayout *mLayout;
+	QHash<QAction *, CustomToolButton *> mActionToButton;
```
Comment (DamirPorobic) [inline src/widgets/Controls.h:49]: This one isn't use either? Can be removed?

File: src/gui/annotator/settings/AnnotationControlsWidget.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+public:
+	AnnotationControlsWidget();
+	~AnnotationControlsWidget() override;
+	void setOrientation(Qt::Orientation orientation) override;
+	QString name() const override;
+
+Q_SIGNALS:
+	void undo() const;
+	void redo() const;
+	void crop() const;
```
Comment (DamirPorobic) [inline src/gui/annotator/settings/AnnotationControlsWidget.h:40]: Just a minor neatpicking thing, the signals `undo` and `redo` perform the real operation, but `crop`, `scale` and `Modify Canvas` actually just show some other windows or widgets without doing the operation yet. Don't know if we should reflect that in the signal names, like `showCrop`, `showScale`? What do you think?

**Review by DamirPorobic — APPROVED** [review]
Looks good to me. The unused variable should be remove and the names are up to you, if you want to rename the signals.

File: src/widgets/TransformationTools.h [hunk 7]
```diff
@@ -45,7 +45,6 @@ Q_OBJECT
 	void modifyCanvas() const;
 
 private:
-	QActionGroup *mActionGroup;
 	QHBoxLayout *mLayout;
 	QHash<QAction *, CustomToolButton *> mActionToButton;
```
[the remark this answers is not in the document]
Comment (precla) [reply src/widgets/TransformationTools.h]: Already removed. That's the old file you're looking at

File: src/gui/annotator/settings/AnnotationControlsWidget.h (hunk 6 above)
Comment (precla) [reply src/gui/annotator/settings/AnnotationControlsWidget.h]: I agree, makes code more understandable. Changed

**Comment by DamirPorobic** [comment]
Thanks for providing this PR, merged.

**Comment by DamirPorobic** [comment]
Run into an issue now that I was implementing support for this in ksnip. This is a setting basically but you can only show it and never hide. Also from the naming, we have `showCrop()`, `showRotator()` (btw this one is missing in the controls, it allows rotating the image) which basically show a temporary window, but this here should be more like an option or setting, do we want to show it or not. We should maybe rename it to something like `isControlsWidgetVisible(bool isVisible)` or `setControlsWidgetEnabled(bool isEnabled)`, dunno, something in that direction. Do you have time to make another PR?

**Comment by precla** [comment]
Oh, sorry for that one. Haven't tested it in ksnip.
You mean rename and change the `showControlsWidget` so it allows to either have it enabled or disabled? Yeah, currently it only allows to never show or show and never hide.. So, following changes in a new PR?

- `showControlsWidget()` -> `setControlsWidgetEnabled(bool isEnabled)`
- add `showRotator()` into `Controls`

**Comment by DamirPorobic** [comment]
Not an issue, you were not supposed to test it in ksnip, I just thought it would be nice to make it there available too if someone want to use it. There it's just a checkbox which is unchecked by default. Not I can enable the control but not disable, I have to restart ksnip xD

Yes, exactly, those two changes. Now looking at those two options, `isControlsWidgetVisible(bool isVisible)` seems to be more accurate because it's visible or not, versus enabled which would be kind of grayed out or not.

Thanks again!

**Comment by precla** [comment]
> Not I can enable the control but not disable, I have to **restart** ksnip xD

The mother of all solutions in IT :D

> Yes, exactly, those two changes. Now looking at those two options, `isControlsWidgetVisible(bool isVisible)` seems to be more accurate because it's visible or not, versus enabled which would be kind of grayed out or not.

I'm on my way to add those :)

**Comment by DamirPorobic** [comment]
yeah :D
image

**Comment by precla** [comment]
:D
btw. started to work on this and just to be sure:
a) `isControlsWidgetVisible(bool isVisible)` has to make the `Controls` appear and disappear? in that case I'd go for `setControlsWidgetVisible(bool setVisible)` because we want to set/unset something
**or**
b) only give feedback if `Controls` is visible: return `true` if currently visible, if not return `false`?

if a) -> we only have [insertDockWidget()](https://github.com/ksnip/kImageAnnotator/blob/master/src/gui/annotator/AnnotationWidget.h#L85) , nothing to hide it. Except [QWidget's hide()](https://doc.qt.io/qt-5/qwidget.html#hide), to hide the `AnnotationDockWidget` created for `Controls`.

**Comment by precla** [comment]
info: Rotate button is in a separate MR: https://github.com/ksnip/kImageAnnotator/pull/264

**Comment by DamirPorobic** [comment]
Definitely `a)` and yes, the naming you have there is correct. 

`QMainWindow` should have a `void removeDockWidget(QDockWidget *dockwidget);` for removing dock widgets, you probably just need to additionally remove it from the `mDockWidgets` List.

**Comment by Pointedstick** [comment]
Nice!

**Pull request merged**